### ISSUE TITLE
Query: GroupJoin with a complex join key fails during compilation

### ISSUE THREAD

**Issue by mbashov** [aspnet/EntityFrameworkCore#11022]
With the following DbContext on EntityFramework Core 2.0.1:
```c#
public class Context : DbContext
{
    public DbSet<Entity> Entities { get; set; }
    public DbSet<Attribute> Attributes { get; set; }
    public DbSet<Value> Values { get; set; }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Value>().HasKey(v => new { v.EntityId, v.AttributeId });
    }
}

public class Entity
{
    public int Id { get; set; }
    public string Name { get; set; }
    public ICollection<Value> Values { get; set; }
}

public class Attribute
{
    public int Id { get; set; }
    public string Name { get; set; }
    public ICollection<Value> Values { get; set; }
}

public class Value
{
    public int EntityId { get; set; }
    public Entity Entity { get; set; }
    public int AttributeId { get; set; }
    public Attribute Attribute { get; set; }
    public int Data { get; set; }
}
```

This query causes exception:
```c#
context
    .Entities
    .Join(context.Attributes, entity => true, attr => true, (entity, attr) => new { entity, attr })
    .GroupJoin(
        context.Values,
        pair => pair.entity.Values.FirstOrDefault(v => v.Attribute == pair.attr),
        value => value,
        (pair, values) => new { pair.entity, pair.attr, value = values.FirstOrDefault() })
    .Select(eav => new { EntityId = eav.entity.Id, AttributeId = eav.attr.Id, Value = (int?)eav.value.Data })
    .ToList();
```

```
System.ArgumentException: 'Expression of type 'System.Func`2[Value,Value]' cannot be used for parameter of type 'System.Func`2[Value,Microsoft.EntityFrameworkCore.Storage.ValueBuffer]' of method 'System.Collections.Generic.IEnumerable`1[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Entity,Microsoft.EntityFrameworkCore.Storage.ValueBuffer],System.Collections.Generic.IEnumerable`1[Value]]] _GroupJoin[TransparentIdentifier`2,Value,ValueBuffer,TransparentIdentifier`2](System.Collections.Generic.IEnumerable`1[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Entity,Microsoft.EntityFrameworkCore.Storage.ValueBuffer]], System.Collections.Generic.IEnumerable`1[Value], System.Func`2[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Entity,Microsoft.EntityFrameworkCore.Storage.ValueBuffer],Microsoft.EntityFrameworkCore.Storage.ValueBuffer], System.Func`2[Value,Microsoft.EntityFrameworkCore.Storage.ValueBuffer], System.Func`3[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Entity,Microsoft.EntityFrameworkCore.Storage.ValueBuffer],System.Collections.Generic.IEnumerable`1[Value],Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[Entity,Microsoft.EntityFrameworkCore.Storage.ValueBuffer],System.Collections.Generic.IEnumerable`1[Value]]])''

Stack trace:
   at System.Dynamic.Utils.ExpressionUtils.ValidateOneArgument(MethodBase method, ExpressionType nodeKind, Expression arguments, ParameterInfo pi, String methodParamName, String argumentParamName, Int32 index)
   at System.Linq.Expressions.Expression.Call(MethodInfo method, Expression arg0, Expression arg1, Expression arg2, Expression arg3, Expression arg4)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitGroupJoinClause(GroupJoinClause groupJoinClause, QueryModel queryModel, Int32 index)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitGroupJoinClause(GroupJoinClause groupJoinClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.Clauses.GroupJoinClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, INodeTypeProvider nodeTypeProvider, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass15_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.Main(String[] args)
```

If the key properties are selected manually, the query is generated as expected:
```c#
context
    .Entities
    .Join(context.Attributes, entity => true, attr => true, (entity, attr) => new { entity, attr })
    .GroupJoin(
        context.Values,
        pair => new { EntityId = pair.entity.Id, AttributeId = pair.attr.Id },
        value => new { value.EntityId, value.AttributeId },
        (pair, values) => new { pair.entity, pair.attr, value = values.FirstOrDefault() })
    .Select(eav => new { EntityId = eav.entity.Id, AttributeId = eav.attr.Id, Value = (int?)eav.value.Data })
    .ToList();
```

```sql
SELECT "entity"."Id" AS "EntityId0", "entity"."Name", "value"."Id", "value"."AttributeId", "value"."Data", "value"."EntityId", "attr"."Id" AS "AttributeId0"
	FROM "Entities" AS "entity"
	INNER JOIN "Attributes" AS "attr" ON TRUE = TRUE
	LEFT JOIN "Values" AS "value" ON ("entity"."Id" = "value"."EntityId") AND ("attr"."Id" = "value"."AttributeId")
	ORDER BY "EntityId0", "AttributeId0"
```

A simpler GroupJoin on navigation property fails with a different exception:
```c#
context
    .Entities
    .GroupJoin(context.Values, entity => entity, value => value.Entity, (e, v) => new { e, v })
    .ToList();
```

```
System.ArgumentException: 'Property 'Entity Entity' is not defined for type 'Entity''

Stack trace:
   at System.Linq.Expressions.Expression.Property(Expression expression, PropertyInfo property)
   at System.Linq.Expressions.Expression.MakeMemberAccess(Expression expression, MemberInfo member)
   at Microsoft.EntityFrameworkCore.Internal.ExpressionExtensions.MakeMemberAccess(Expression expression, MemberInfo member)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.<>c__DisplayClass18_0.<VisitMember>b__1(Expression e)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.CreateSubqueryForNavigations(Expression outerQuerySourceReferenceExpression, ICollection`1 navigations, Func`2 propertyCreator)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.RewriteNavigationProperties(IReadOnlyList`1 properties, IQuerySource querySource, Expression expression, Expression declaringExpression, String propertyName, Type propertyType, Func`2 propertyCreator, Func`2 conditionalAccessPropertyCreator)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.<>c__DisplayClass18_0.<VisitMember>b__0(IReadOnlyList`1 ps, IQuerySource qs)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindPropertyExpressionCore[TResult](Expression propertyExpression, IQuerySource querySource, Func`3 propertyBinder)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindNavigationPathPropertyExpression[TResult](Expression propertyExpression, Func`3 propertyBinder)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitMember(MemberExpression node)
   at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.NavigationRewritingQueryModelVisitor.VisitJoinClauseInternal(JoinClause joinClause)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.NavigationRewritingQueryModelVisitor.VisitJoinClause(JoinClause joinClause, QueryModel queryModel, GroupJoinClause groupJoinClause)
   at Remotion.Linq.Clauses.JoinClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, GroupJoinClause groupJoinClause)
   at Remotion.Linq.QueryModelVisitorBase.VisitGroupJoinClause(GroupJoinClause groupJoinClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.Clauses.GroupJoinClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.Rewrite(QueryModel queryModel, QueryModel parentQueryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel, Boolean asyncQuery)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel, Boolean asyncQuery)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, INodeTypeProvider nodeTypeProvider, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass15_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.Main(String[] args)
```

**Comment by maumar** [comment]
In case key selectors for Join/GroupJoin are entity, we should rewrite it to key comparisons, like we do for equality

**Comment by smitpatel** [comment]
This would probably get covered in entityequality refactoring fix. You can assign to me.

**Comment by maumar** [comment]
fixed in fdb8eceb74a3fd585f90eced744977f74477b1dd
